### PR TITLE
rpc: Add missing cs_main lock in getblocktemplate(...)

### PR DESCRIPTION
**Pull request by practicalswift**
Reading the variable `chainActive` requires holding the mutex `cs_main`.

Prior to this commit the `cs_main` mutex was not held when accessing `chainActive` in:

```
while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
```

### CODE REVIEW

**Comment by practicalswift** [comment]
@luke-jr @promag, would you mind reviewing this revised version? :-)

File: src/rpc/mining.cpp [hunk 1]
```diff
@@ -479,7 +479,12 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
             checktxtime = std::chrono::steady_clock::now() + std::chrono::minutes(1);
 
             WaitableLock lock(csBestBlock);
-            while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
+            bool keepRunning;
+            {
+                LOCK(cs_main);
+                keepRunning = chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning();
+            }
+            while (keepRunning)
```
Comment (promag) [inline src/rpc/mining.cpp]: My suggestion is to create a function like:
```cpp
bool KeepRunning() {
    LOCK(cs_main);
    assert(chainActive.Tip());
    return chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning();
}
```
And here just:
```cpp
while (KeepRunning()) {
    ...
}
```

**Review by promag — COMMENTED** [review]
Concept ACK.

**Comment by promag** [comment]
Restarted travis job.

**Comment by luke-jr** [comment]
The whole point of csBestBlock was that we shouldn't need cs_main to simply check the best block. It looks like the logic for that got lost somewhere, and I'm not sure we're even safe from races anymore. :(

Suggest instead of this, fixing csBestBlock so it is sufficient/works.

**Comment by practicalswift** [comment]
@promag Now using lambda - please re-review :-)

**Comment by practicalswift** [comment]
@luke-jr Would fixing this make things worse?

**Comment by practicalswift** [comment]
Pinging @luke-jr :-)

**Comment by TheBlueMatt** [comment]
No reason to wait on a "better fix" before merging this. Lets just merge and then @luke-jr can fix it in the way he wants in a cleanup follow-up.

utACK 6e65f35f23337cf0b284acb2bb2ec038ae83aeca

**Comment by theuni** [comment]
I agree with @luke-jr on this one. I'd rather fix it to use csBestBlock properly than waste the time figuring out the possible impact of locking cs_main in the loop.

**Comment by TheBlueMatt** [comment]
@theuni i find the current code relatively readable, and I'm really not at all sure what "fixing csBestBlock" means, given a condition variable is allowed to wake up spuriously whenever it wants.

**Comment by theuni** [comment]
@TheBlueMatt as @luke-jr mentioned, presumably cvBlockChange was at one point paired with a uint256 reflecting the miner's best known hash, which at some point got removed. We shouldn't need to take cs_main to check whether the new tip matches the cached one.

**Comment by practicalswift** [comment]
If the scope of this grows beyond the initial goal of fixing the missing `cs_main` lock I'll have to leave this for others to fix. Let me know when a fix for the issue raised by @luke-jr is merged and I'll close this PR.

If that is something that is unlikely to happen soon I'd vote for merging this fix while waiting for a future better fix.

**Comment by TheBlueMatt** [comment]
@theuni "fixing" cvBlockChanged/csBestBlock probably means wholesale removing it and replacing it with appropriate event listeners, but for that, I think, we (realistically) need to beef up the infrastructure behind CValidationInterface so that we dont end up with wallet or other things blocking each other just because they're all on the validation interface (or add some new interface for this stuff that isnt "call into RPC stuff from inside ConnectBlock"), but unless you want to do that, I really think it shouldnt block a simple lock fix.

**Comment by theuni** [comment]
Ok, since we re-lock cs_main before returning anyway, and since cs_main being locked for any notable period of time after a legitimate cvBlockChange notification would likely indicate a rapid 2nd block anyway, I suppose no regressions would be introduced.

So, -0 to 6e65f35f23337cf0b284acb2bb2ec038ae83aeca. Looks correct to fix the unguarded access.

**Comment by practicalswift** [comment]
Any chance of getting this merged? :-)

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -479,7 +479,12 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
             checktxtime = std::chrono::steady_clock::now() + std::chrono::minutes(1);
 
             WaitableLock lock(csBestBlock);
-            while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
+            auto KeepRunning = [&]() -> bool {
+                LOCK(cs_main);
+                assert(chainActive.Tip());
+                return chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning();
+            };
+            while (KeepRunning())
```
Comment (promag) [inline src/rpc/mining.cpp]: Sorry but how about plain code?
```cpp
while (IsRPCRunning()) {
    {
        LOCK(cs_main);
        if (chainActive.Tip()->GetBlockHash() != hashWatchedChain) break;
    }
    ...
```

File: src/rpc/mining.cpp [hunk 3]
```diff
@@ -479,7 +479,12 @@ UniValue getblocktemplate(const JSONRPCRequest& request)
             checktxtime = std::chrono::steady_clock::now() + std::chrono::minutes(1);
 
             WaitableLock lock(csBestBlock);
-            while (chainActive.Tip()->GetBlockHash() == hashWatchedChain && IsRPCRunning())
+            auto KeepRunning = [&]() -> bool {
+                LOCK(cs_main);
+                assert(chainActive.Tip());
```
Comment (promag) [inline src/rpc/mining.cpp]: I would drop the assert, otherwise should we add all over where tip is used?

**Review by promag — COMMENTED** [review]
utACK 6e65f35, but I have 2 comments.

**Comment by practicalswift** [comment]
@promag Good suggestion. Please re-review!

@TheBlueMatt @theuni Would you mind re-reviewing in light of the latest commit? :-)

**Comment by sipa** [comment]
Here is an alternative: #12743

**Comment by practicalswift** [comment]
Closing in favour of #12743 which is a better alternative :-)

**Pull request closed without merging**